### PR TITLE
[added] description of nats based account resolver

### PR DESCRIPTION
**Pull request by matthiashanel**
Signed-off-by: Matthias Hanel <mh@synadia.com>

### CODE REVIEW

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 1]
```diff
@@ -1,9 +1,10 @@
 # Account lookup using Resolver
 
-The `resolver` configuration option is used in conjunction with [NATS JWT Authentication](./) and [nsc](../../../../nats-tools/nsc/). The `resolver` option specifies a URL where the nats-server can retrieve an account JWT. There are two built-in resolver implementations:
+The `resolver` configuration option is used in conjunction with [NATS JWT Authentication](./) and [nsc](../../../../nats-tools/nsc/). The `resolver` option specifies a URL where the nats-server can retrieve an account JWT. There are three built-in resolver implementations:
 
-* `URL`
-* `MEMORY`
+* [`URL`](resolver.md#URL-Resolver)
+* [`MEMORY`](resolver.md#Memory)
+* [nats based resolver](resolver.md#nats-based-resolver)
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:7]: Is this the literal string specified?

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 2]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
 
 For more information on how to configure a memory resolver, see [this tutorial](mem_resolver.md).
 
+## nats based resolver
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:38]: Capitalize NATS

File: nats-server/configuration/securing_nats/jwt/resolver.md (hunk 1 above)
Comment (matthiashanel) [reply nats-server/configuration/securing_nats/jwt/resolver.md:7]: no it is not, we use a struct instead

File: nats-server/configuration/securing_nats/jwt/resolver.md (hunk 2 above)
Comment (matthiashanel) [reply nats-server/configuration/securing_nats/jwt/resolver.md]: fixed

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 3]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
 
 For more information on how to configure a memory resolver, see [this tutorial](mem_resolver.md).
 
+## NATS Based Resolver
+
+NATS based resolver embed the functionality of the [account server](https://github.com/nats-io/nats-account-server) inside the nats-server.
+To not have to store all account jwt on every server, this resolver has two sub types `full` and `cache`.
+Their commonalities are that they exchange/lookup account jwt via NATS and the system account and store them in a local (not shared) directory.
+
+### full
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:44]:
```suggestion
### Full
```

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 4]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
...
+    # Interval at which a nats-server with a nats based account resolver will compare
+    # it's state with one random nats based account resolver in the cluster and if needed,
+    # exchange jwt and converge on the same set of jwt.
+    interval: "2m"
+    # limit on the number of jwt stored, will reject new jwt once limit is hit.
+    limit: 1000
+}
+```
+
+This resolver type also supports `resolver_preload`. When present jwt listed are stored in the resolver.
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:70]:
```suggestion
This resolver type also supports `resolver_preload`. When present JWT's are listed are stored in the resolver.
```

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 5]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
...
+There, they may be subject to updates. Restarts of the `nats-server` will hold on to these more recent versions.
+
+Not every server in a cluster needs to be set to `full`.
+You need enough to still serve your workload adequately, while some server are offline.
+
+### cache
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:76]:
```suggestion
### Cache
```

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 6]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
...
+
+This resolver only stores a subset of [JWT](../../nats-server/configuration/securing_nats/jwt/) and evicts extra ones based on an LRU scheme.
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:78]:
```suggestion
This resolver only stores a subset of [JWTs](../../nats-server/configuration/securing_nats/jwt/) and evicts others based on an LRU scheme. 
```

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 7]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
...
+    # Directory in which account jwt will be store
+    dir: "./"
+    # limit on the number of jwt stored, will evict old jwt once limit is hit.
+    limit: 1000
+    # How long to hold on to a jwt before discarding it. 
+    ttl: "2m"
+}
+```
+
+### nats based resolver - integration
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:94]:
```suggestion
### NATS Based Resolver - Integration
```

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 8]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
...
+
+nats based resolver utilize the system account for lookup and upload of account [JWTs](../../nats-server/configuration/securing_nats/jwt/) .
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:96]:
```suggestion
The NATS based resolver utilizes the system account for lookup and upload of account [JWTs](../../nats-server/configuration/securing_nats/jwt/) .
```

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 9]
```diff
@@ -34,3 +35,68 @@ The `MEMORY` resolver is recommended when the server has a small number of accou
...
+If your application requires tighter integration you can make use of these subjects for tighter integration.
+
+To upload or update a possibly on the fly generated account jwt without [`nsc`](../../../../nats-tools/nsc/README.md), send it as request to `$SYS.REQ.CLAIMS.UPDATE`.
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:99]:
```suggestion
To upload or update any generated account jwt without [`nsc`](../../../../nats-tools/nsc/README.md), send it as request to `$SYS.REQ.CLAIMS.UPDATE`.
```

File: nats-tools/nas/README.md [hunk 10]
```diff
@@ -1,11 +1,14 @@
...
-
-* a [directory](nas_conf.md#directory-configuration)
-* an [NSC](../nsc/nsc.md) [directory](nas_conf.md#nsc-configuration)
+The [NATS Account Server](https://github.com/nats-io/nats-account-server) is an HTTP server that hosts and vends [JWTs](../../nats-server/configuration/securing_nats/jwt/) for nats-server 2.0 account authentication. The server supports an number of stores which enable it to serve account [JWTs](../../nats-server/configuration/securing_nats/jwt/) from a [directory](nas_conf.md#directory-configuration)
 
 > The nats server can be configured with a [memory resolver](../../nats-server/configuration/securing_nats/jwt/resolver.md#memory) as well. This avoids usage of the account server.
+> The nats server can be configured with a [nats based resolver](../../nats-server/configuration/securing_nats/jwt/resolver.md#nats-based-resolver) for the same purpose as well.
```
Comment (ColinSullivan1) [inline nats-tools/nas/README.md:6]:
```suggestion
> The NATS server can be configured with a [NATS based resolver](../../nats-server/configuration/securing_nats/jwt/resolver.md#nats-based-resolver) for the same purpose as well.
```

File: nats-tools/nas/README.md [hunk 11]
```diff
@@ -1,11 +1,14 @@
...
+> 
+> Usage of [full nats based resolver](../../nats-server/configuration/securing_nats/jwt/resolver.md#nats-based-resolver) over [NATS Account Server](https://github.com/nats-io/nats-account-server) is recommended.
```
Comment (ColinSullivan1) [inline nats-tools/nas/README.md:8]:
```suggestion
> Usage of [full NATS based resolver](../../nats-server/configuration/securing_nats/jwt/resolver.md#nats-based-resolver) over [NATS Account Server](https://github.com/nats-io/nats-account-server) is recommended.
```

File: nats-server/configuration/securing_nats/jwt/resolver.md [hunk 12]
```diff
@@ -1,9 +1,10 @@
 # Account lookup using Resolver
```
Comment (ColinSullivan1) [inline nats-server/configuration/securing_nats/jwt/resolver.md:1]:
```suggestion
# Account Lookup Using a Resolver
```

**Comment by ColinSullivan1** [comment]
Thanks @matthiashanel , I added some suggestions.

File: nats-server/configuration/securing_nats/jwt/resolver.md (hunk 2 above)
Comment (matthiashanel) [reply nats-server/configuration/securing_nats/jwt/resolver.md]: fixed

File: nats-server/configuration/securing_nats/jwt/resolver.md (hunk 3 above)
Comment (matthiashanel) [reply nats-server/configuration/securing_nats/jwt/resolver.md:44]: fixed

**Comment by matthiashanel** [comment]
@ColinSullivan1 thanks for your suggestions, I accepted all of them.

**Review by ColinSullivan1 — APPROVED** [review]
LGTM.

**Comment by gcolliso** [comment]
lgtm

**Pull request merged**